File: backends/js/src/utils/stake.ts
import { StakeAccount, ACCESS_PROGRAM_ID, StakePool, BondAccount, getBondAccounts } from "@access-protocol";
import { connection } from "./connection";
import { PublicKey } from "@solana/web3.js";
import BN from 'bn.js';

/**
 * Public key of the stake pool (must be defined in the .env file)
 */
const STAKE_POOL_KEY = new PublicKey(process.env.STAKE_POOL_KEY!);

/**
 * Verifies that a user has enough tokens staked
 * @param owner The owner of the stake account
 * @returns
 */
export const checkStake = async (owner: string) => {
  const [key] = await StakeAccount.getKey(
    ACCESS_PROGRAM_ID,
    new PublicKey(owner),
    STAKE_POOL_KEY
  );
  const stakeAccount = await StakeAccount.retrieve(connection, key);
  const stakePool = await StakePool.retrieve(
    connection,
    stakeAccount.stakePool
  );

  const bondAccounts = await getBondAccounts(connection, new PublicKey(owner));
  const bondTotalStaked = bondAccounts
    .map((bondAccount): BN => {
      const account = BondAccount.deserialize(bondAccount.account.data);
      if (account.stakePool.toBase58() === STAKE_POOL_KEY.toBase58()) {
        return account.totalStaked;
      }
      
      return new BN(0);
    })
    .reduce((totalStaked, stakeAmount) => {
      stakeAmount.add(totalStaked);
    }, new BN(0));


  const requiredAmount = Math.min(
    stakeAccount.poolMinimumAtCreation.toNumber(),
    stakePool.minimumStakeAmount.toNumber()
  );

  return stakeAccount.stakeAmount.add(bondTotalStaked).toNumber() > requiredAmount;
  };

File: backends/go/utils/stake.go
package utils

import (
	"context"

	"encoding/binary"

	"os"

	"github.com/portto/solana-go-sdk/client"
	"github.com/portto/solana-go-sdk/common"
	"github.com/portto/solana-go-sdk/rpc"
)

const OFFSET_STAKE = 1 + 32
const OFFSET_MIN_CREATION = 1 + 32 + 8 + 32 + 8
const OFFSET_MIN_CURRENT = 1 + 1 + 2 + 4 


// Verifies that a user has enough tokens staked
func CheckStake(owner string) bool {
	c := client.NewClient(rpc.DevnetRPCEndpoint)

	
	seeds := [][]byte{	
		[]byte("stake_account"), 
		common.PublicKeyFromString(owner).Bytes(), 
		common.PublicKeyFromString(os.Getenv("STAKE_POOL_KEY")).Bytes(),
	}
	key, _, err := common.FindProgramAddress(seeds, common.PublicKeyFromString(os.Getenv("STAKE_POOL_KEY")))

	if err != nil {
		// Error deriving key
		return false
	}

	info, err := c.GetAccountInfo(context.TODO(), key.ToBase58())

	if err != nil {
		// Stake account not found
		return false
	}

	stake := uint64(binary.LittleEndian.Uint64(info.Data[OFFSET_STAKE: OFFSET_STAKE + 8]))
	min := uint64(binary.LittleEndian.Uint64(info.Data[OFFSET_MIN_CREATION: OFFSET_MIN_CREATION + 8]))


	info, err = c.GetAccountInfo(context.TODO(), os.Getenv("STAKE_POOL_KEY"))

	if err != nil {
		// Stake pool not found
		return false
	}

	pool_minimum := uint64(binary.LittleEndian.Uint64(info.Data[OFFSET_MIN_CURRENT: OFFSET_MIN_CURRENT + 8]))

	return stake > min || stake > pool_minimum
}